spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/MergeIntoExec.scala

```diff
@@ -0,0 +1,111 @@
+import org.apache.spark.sql.catalyst.expressions.{Attribute, BasePredicate, Expression, UnsafeProjection}
+import org.apache.spark.sql.catalyst.expressions.codegen.GeneratePredicate
+import org.apache.spark.sql.catalyst.plans.logical.MergeIntoParams
+import org.apache.spark.sql.execution.{SparkPlan, UnaryExecNode}
+
+case class MergeIntoExec(mergeIntoProcessor: MergeIntoParams,
+                         @transient targetRelation: DataSourceV2Relation,
+                         override val child: SparkPlan) extends UnaryExecNode {
+
+  override def output: Seq[Attribute] = targetRelation.output
+
+  protected override def doExecute(): RDD[InternalRow] = {
+    child.execute().mapPartitions {
+      processPartition(mergeIntoProcessor, _)
+    }
+  }
+
+  private def generateProjection(exprs: Seq[Expression], attrs: Seq[Attribute]): UnsafeProjection = {
+    UnsafeProjection.create(exprs, attrs)
+  }
+
+  private def generatePredicate(expr: Expression, attrs: Seq[Attribute]): BasePredicate = {
+    GeneratePredicate.generate(expr, attrs)
+  }
+
+  def applyProjection(predicates: Seq[BasePredicate],
+                      projections: Seq[UnsafeProjection],
+                      projectTargetCols: UnsafeProjection,
+                      projectDeleteRow: UnsafeProjection,
+                      inputRow: InternalRow,
+                      targetRowNotPresent: Boolean): InternalRow = {
+    // Find the first combination where the predicate evaluates to true
+    val pair = (predicates zip projections).find {
+      case (predicate, _) => predicate.eval(inputRow)
+    }
+
+    // Now apply the appropriate projection to either :
+    // - Insert a row into target
+    // - Update a row of target
+    // - Delete a row in target. The projected row will have the deleted bit set.
+    pair match {
+      case Some((_, projection)) =>
+        projection.apply(inputRow)
+      case None =>
+        if (targetRowNotPresent) {
+          projectDeleteRow.apply(inputRow)
+        } else {
+          projectTargetCols.apply(inputRow)
+        }
+    }
+  }
+
+  def processPartition(params: MergeIntoParams,
+                       rowIterator: Iterator[InternalRow]): Iterator[InternalRow] = {
+    val joinedAttrs = params.joinedAttributes
+    val isSourceRowNotPresentPred = generatePredicate(params.isSourceRowNotPresent, joinedAttrs)
+    val isTargetRowNotPresentPred = generatePredicate(params.isTargetRowNotPresent, joinedAttrs)
+    val matchedPreds = params.matchedConditions.map(generatePredicate(_, joinedAttrs))
+    val matchedProjs = params.matchedOutputs.map(generateProjection(_, joinedAttrs))
+    val notMatchedPreds = params.notMatchedConditions.map(generatePredicate(_, joinedAttrs))
+    val notMatchedProjs = params.notMatchedOutputs.map(generateProjection(_, joinedAttrs))
+    val projectTargetCols = generateProjection(params.targetOutput, joinedAttrs)
+    val projectDeletedRow = generateProjection(params.deleteOutput, joinedAttrs)
```

Comment:
These last two projections are only needed when `notMatchedPreds` or `matchedPreds` does not have a default case, i.e. `lit(true)`.

In the rewrite, there is also a function, `getClauseCondition`, that fills in `lit(true)` if there is no clause condition. But I don't think that any predicates _after_ the `true` condition are dropped.

I think we could simplify the logic here and avoid extra clauses by ensuring that both `matchedPreds` and `notMatchedPreds` end with `lit(true)`. Then this class would not need to account for the case where no predicate matches and we wouldn't have extra predicates passed through.

Reply:
I guess `projectTargetCols` would still be needed for the case where the source row isn't present, but it would still make this a little simpler. Especially the `applyProjection` method.

Reply:
@rdblue 

1. ```projectTargetCols``` represents the expression that needs to be applied on the output of outer join which has columns from both the tables to only project the target output columns plus the deleted flag set to false.
2. ```projectDeletedRow ``` does the same but with the "deleted flag". I think in the earlier comment we discussed possible ideas to optimize this (will address in follow-up)
3. ``matchedPreds``` and ```notMatchedPred```  go hand in hand with their corresponding projections that is specified by the user in the `WHEN MATCHED ` and `WHEN NOT MATCHED` clauses.

Given this background, can you please explain your idea a little bit ?

Reply:
Let's clear this up in a follow-up.